### PR TITLE
local date macro

### PR DESCRIPTION
**Pull request by ilaflamme-va**

### CODE REVIEW

File: charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java [hunk 1]
```diff
@@ -26,6 +28,14 @@ ConnectionDetails asConnectionDetails(String name, String value) {
     if (parts.length != 4) {
       throw badValue(name, value, "incorrect number of parts");
     }
+    try {
+      ZoneId.of(parts[3]);
+    } catch (ZoneRulesException e) {
+      throw badValue(
+          name,
+          value,
+          "vistalink.properties: vista-connection string is configured with a bad timezone");
```
Comment (bschofield-va) [inline charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java:37]: Update the error message with something helpful. This currently says "There is a bad value somewhere in the file. Good luck!" We know the exactly what entry is bad, let's include that in the error message.

File: charon/src/main/java/gov/va/api/lighthouse/charon/service/controller/MacroProcessorFactory.java [hunk 2]
```diff
@@ -14,7 +15,12 @@
 
   @NonNull List<Macro> macros;
 
-  public MacroProcessor create(MacroExecutionContext ctx) {
-    return MacroProcessor.builder().macros(macros).macroExecutionContext(ctx).build();
+  /** Create a macro processor factory from a context and connectionDetails. */
+  public MacroProcessor create(MacroExecutionContext ctx, ConnectionDetails connectionDetails) {
```
Comment (bschofield-va) [inline charon/src/main/java/gov/va/api/lighthouse/charon/service/controller/MacroProcessorFactory.java:19]: The MacroExecutionContext should have _everything_ we need to evaluate macros. Remove this parameter, update the context.

File: charon/src/main/java/gov/va/api/lighthouse/charon/service/controller/MacroProcessor.java [hunk 3]
```diff
@@ -20,6 +21,8 @@
 
   @NonNull MacroExecutionContext macroExecutionContext;
 
+  @NonNull ConnectionDetails connectionDetails;
```
Comment (bschofield-va) [inline charon/src/main/java/gov/va/api/lighthouse/charon/service/controller/MacroProcessor.java:24]: remove

File: charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java (hunk 1 above)
Comment (ilaflamme-va) [reply charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java:37]: It gives a nice one. 
```java.lang.IllegalArgumentException: Cannot parse value n="host:1234:div:Space/Moon", expected "hostname:port:divisionIen:timezone" (vistalink.properties: vista-connection string is configured with a bad timezone)```

File: charon/pom.xml [hunk 4]
```diff
@@ -39,6 +39,11 @@
       <artifactId>charon-api</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>gov.va.api.lighthouse</groupId>
+      <artifactId>charon-models</artifactId>
```
Comment (bschofield-va) [inline charon/pom.xml:44]: I'm not crazy about this dependency. I don't really like Charon having class-level knowledge the different types of models. I wonder if we should move `FilemanDate`. The `api` library doesn't feel like a great home for that, but not sure where else it would go. Thoughts?

File: charon/requests/local-date-macro-vpr-get-patient-data-xml.json [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  "target": {
+    "forPatient":"ignored-for-now"
+  },
+  "rpc": {
+    "name":"VPR GET PATIENT DATA",
+    "context":"VPR APPLICATION PROXY",
+    "parameters": [
+      {"string":";195607"},
+      {"string":"demographics;vitals"},
+      {"string":"${local-fileman-date(2021-02-22T13:28:00Z)"},
```
Comment (bschofield-va) [inline charon/requests/local-date-macro-vpr-get-patient-data-xml.json:15]: ^ that doesn't look like it will work... it's missing a close `}`

File: charon/src/main/java/gov/va/api/lighthouse/charon/service/controller/LocalDateMacro.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+public class LocalDateMacro implements Macro {
+
+  @Override
+  public String evaluate(MacroExecutionContext ctx, String value) {
+    try {
+      return FilemanDate.from(Instant.parse(value))
+          .formatAsDateTime(ZoneId.of(ctx.connectionDetails().timezone()));
+    } catch (DateTimeParseException e) {
+      throw new LocalDateMacroParseFailure();
```
Comment (bschofield-va) [inline charon/src/main/java/gov/va/api/lighthouse/charon/service/controller/LocalDateMacro.java:16]: Is there something we can do here to provide ourselves some clue as to why/where this failed? Can we not parse  ISO-8601 date, can we not format it?

File: charon/src/main/java/gov/va/api/lighthouse/charon/service/controller/VistalinkRpcInvoker.java [hunk 7]
```diff
@@ -88,6 +87,10 @@ public void close() {
     }
   }
 
+  public ConnectionDetails connectionDetails() {
```
Comment (bschofield-va) [inline charon/src/main/java/gov/va/api/lighthouse/charon/service/controller/VistalinkRpcInvoker.java:90]: delete this and add `@Getter` to `connectionDetails` property

File: charon/src/main/java/gov/va/api/lighthouse/charon/service/controller/LocalDateMacro.java (hunk 6 above)
Comment (ilaflamme-va) [reply charon/src/main/java/gov/va/api/lighthouse/charon/service/controller/LocalDateMacro.java:16]: Yea. I could split this up to be more robust on fail.

File: charon/pom.xml (hunk 4 above)
Comment (ilaflamme-va) [reply charon/pom.xml:44]: I'm fine with moving it. API would work but feels wrong.

File: charon/requests/local-date-macro-vpr-get-patient-data-xml.json (hunk 5 above)
Comment (ilaflamme-va) [reply charon/requests/local-date-macro-vpr-get-patient-data-xml.json:15]: Interesting, the request doesn't fail...

Comment (ilaflamme-va) [reply charon/requests/local-date-macro-vpr-get-patient-data-xml.json]: Fixed this bug... the sample request will now get 4 of the 25 total vitals via a search.

**Comment by ilaflamme-va** [comment]
There is now a sample request that returns a subset of vitals based on fileman date.

File: charon/src/test/java/gov/va/api/lighthouse/charon/service/controller/LocalDateMacroTest.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class LocalDateMacroTest {
+  @Mock MacroExecutionContext ctx;
+
+  @Test
+  void evaluateParsesGoodDates() {
```
Comment (jhulbert-va) [inline charon/src/test/java/gov/va/api/lighthouse/charon/service/controller/LocalDateMacroTest.java:18]: Should this check a second time-zone?

File: charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java [hunk 9]
```diff
@@ -26,6 +28,14 @@ ConnectionDetails asConnectionDetails(String name, String value) {
     if (parts.length != 4) {
       throw badValue(name, value, "incorrect number of parts");
     }
+    try {
+      ZoneId.of(parts[3]);
```
Comment (jhulbert-va) [inline charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java:32]: Would it be beneficial to just update connection details so the timezone is a ZoneId, not a String?

File: charon/src/test/java/gov/va/api/lighthouse/charon/service/controller/LocalDateMacroTest.java (hunk 8 above)
Comment (ilaflamme-va) [reply charon/src/test/java/gov/va/api/lighthouse/charon/service/controller/LocalDateMacroTest.java:18]: sure.

File: charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java (hunk 9 above)
Comment (ilaflamme-va) [reply charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java:32]: This logic would just move. At this point, I don't think I care enough.

Comment (jhulbert-va) [reply charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java:32]: It would avoid having to parse a second time as well.

**Review by jhulbert-va — DISMISSED** [review]

File: charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java (hunk 9 above)
Comment (ilaflamme-va) [reply charon/src/main/java/gov/va/api/lighthouse/charon/service/config/VistalinkPropertiesConfig.java:32]: We want the validation of the property to happen when `vistalink.properties` is loaded. ASAP.

File: charon/pom.xml (hunk 4 above)
Comment (ilaflamme-va) [reply charon/pom.xml:43]: Perhaps it just moves to `gov.va.api.lighthouse.charon` ?

File: charon/src/test/java/gov/va/api/lighthouse/charon/service/controller/InteractiveTestSupport.java [hunk 10]
```diff
@@ -25,7 +25,7 @@ static ConnectionDetails localTampaConnectionDetails() {
   }
 
   static MacroExecutionContext nullMacroExecutionContext() {
-    return request -> null;
+    return null;
```
Comment (bschofield-va) [inline charon/src/test/java/gov/va/api/lighthouse/charon/service/controller/InteractiveTestSupport.java:28]: Why this change.... before it was a macro execution context that _returned null_, not it is a _null context itself_?

**Review by bschofield-va — APPROVED** [review]
